crates/ruff_python_semantic/src/red_knot/types.rs

```diff
@@ -0,0 +1,705 @@
+    use ruff_db::vfs::system_path_to_file;
+
+    use crate::db::tests::TestDb;
+    use crate::module::resolver::{set_module_resolution_settings, ModuleResolutionSettings};
+    use crate::red_knot::semantic_index::root_scope;
+    use crate::red_knot::types::{
+        expression_ty, infer_types, public_symbol_ty_by_name, TypingContext,
+    };
+
+    fn setup_db() -> TestDb {
+        let mut db = TestDb::new();
+        set_module_resolution_settings(
+            &mut db,
+            ModuleResolutionSettings {
+                extra_paths: vec![],
+                workspace_root: FileSystemPathBuf::from("/src"),
+                site_packages: None,
+                custom_typeshed: None,
+            },
+        );
+
+        db
+    }
+
+    #[test]
+    fn local_inference() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_file("/src/a.py", "x = 10")?;
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        let parsed = parsed_module(&db, a);
+
+        let statement = parsed.suite().first().unwrap().as_assign_stmt().unwrap();
+
+        let literal_ty = expression_ty(&db, a, &statement.value);
+
+        assert_eq!(
+            format!("{}", literal_ty.display(&TypingContext::global(&db))),
+            "Literal[10]"
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn dependency_public_symbol_type_change() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.memory_file_system().write_files([
+            ("/src/a.py", "from foo import x"),
+            ("/src/foo.py", "x = 10\ndef foo(): ..."),
+        ])?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+        let x_ty = public_symbol_ty_by_name(&db, a, "x").unwrap();
+
+        assert_eq!(
+            x_ty.display(&TypingContext::global(&db)).to_string(),
+            "Literal[10]"
+        );
+
+        // Change `x` to a different value
+        db.memory_file_system()
+            .write_file("/src/foo.py", "x = 20\ndef foo(): ...")?;
+
+        let foo = system_path_to_file(&db, "/src/foo.py").unwrap();
+        foo.touch(&mut db);
+        std::thread::sleep(Duration::from_millis(1));
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        db.clear_salsa_events();
+        let x_ty_2 = public_symbol_ty_by_name(&db, a, "x").unwrap();
+
+        assert_eq!(
+            x_ty_2.display(&TypingContext::global(&db)).to_string(),
+            "Literal[20]"
+        );
+
+        let a_root_scope = root_scope(&db, a);
+        let infer_types_ingredient_index = infer_types_ingredient_index(&db);
+        let events = db.take_sale_events();
+
+        assert!(events.iter().any(|event| {
+            if let salsa::EventKind::WillExecute { database_key } = &event.kind {
+                database_key.ingredient_index() == infer_types_ingredient_index
+                    && database_key.key_index() == a_root_scope.as_id()
+            } else {
```

Comment:
> It would be nice to have some helpers so that it is less verbose and complicated to make an assertion like "exactly these scopes had to re-run inference". I anticipate we'll be doing this in tests a lot.

I'm not sure how often we'll write these kinds of tests. Ideally, we only have a few tests for this. 

I think what I would like to have is a nice way of printing the query execution plan from salsa and snapshoting it. It's slightly less explicit but it also has the benefit that adding or removing intermediate query shows how that effects the systems caching and invalidation behavior.

Reply:
Hmm, I don't agree with the snapshot approach. I find snapshot tests pretty hard to understand and manage, and way too easy to wrongly update the snapshot without even realizing that it's invalidating the entire point of the test, or rendering the test itself wrong. 

I think it results in better tests if we have nicely abstracted APIs to be used by human-written tests, designed for making very specific and easy to read and understand assertions at a higher level of abstraction about specific conditions that we care about, eg "when we make this change to module y and re-query a type in module z, we don't have to re-infer types in module x"

Reply:
> I think it results in better tests if we have nicely abstracted APIs to be used by human-written tests

I like that too but I struggled designing this with all the generics involved with Salsa. We might be able to come up with a macro. Anyway, I don't think that's something we need solving now.

Reply:
You might think differnetly of snapshot tests if the query execution plan is a nice mermaid flow chart ;)

Reply:
So much pain

```rust
pub(crate) fn assert_will_run_function_query<C, Db, Jar>(
        db: &Db,
        to_function: impl FnOnce(&C) -> &salsa::function::FunctionIngredient<C>,
        key: C::Key,
        events: &[salsa::Event],
    ) where
        C: salsa::function::Configuration<Jar = Jar>
            + salsa::storage::IngredientsFor<Jar = Jar, Ingredients = C>,
        Jar: HasIngredientsFor<C>,
        Db: salsa::DbWithJar<Jar>,
        C::Key: AsId,
    {
        will_run_function_query(db, to_function, key, events, true);
    }

    pub(crate) fn assert_will_not_run_function_query<C, Db, Jar>(
        db: &Db,
        to_function: impl FnOnce(&C) -> &salsa::function::FunctionIngredient<C>,
        key: C::Key,
        events: &[salsa::Event],
    ) where
        C: salsa::function::Configuration<Jar = Jar>
            + salsa::storage::IngredientsFor<Jar = Jar, Ingredients = C>,
        Jar: HasIngredientsFor<C>,
        Db: salsa::DbWithJar<Jar>,
        C::Key: AsId,
    {
        will_run_function_query(db, to_function, key, events, false);
    }

    fn will_run_function_query<C, Db, Jar>(
        db: &Db,
        to_function: impl FnOnce(&C) -> &salsa::function::FunctionIngredient<C>,
        key: C::Key,
        events: &[salsa::Event],
        should_run: bool,
    ) where
        C: salsa::function::Configuration<Jar = Jar>
            + salsa::storage::IngredientsFor<Jar = Jar, Ingredients = C>,
        Jar: HasIngredientsFor<C>,
        Db: salsa::DbWithJar<Jar>,
        C::Key: AsId,
    {
        let (jar, _) =
            <_ as salsa::storage::HasJar<<C as salsa::storage::IngredientsFor>::Jar>>::jar(db);
        let ingredient = jar.ingredient();

        let function_ingredient = to_function(ingredient);

        let ingredient_index =
            <salsa::function::FunctionIngredient<C> as Ingredient<Db>>::ingredient_index(
                function_ingredient,
            );

        let did_run = events.iter().any(|event| {
            if let salsa::EventKind::WillExecute { database_key } = event.kind {
                database_key.ingredient_index() == ingredient_index
                    && database_key.key_index() == key.as_id()
            } else {
                false
            }
        });

        if should_run && !did_run {
            panic!(
                "Expected query {:?} to run but it didn't",
                DebugIdx {
                    db: PhantomData::<Db>,
                    value_id: key.as_id(),
                    ingredient: function_ingredient,
                }
            );
        } else if !should_run && did_run {
            panic!(
                "Expected query {:?} not to run but it did",
                DebugIdx {
                    db: PhantomData::<Db>,
                    value_id: key.as_id(),
                    ingredient: function_ingredient,
                }
            );
        }
    }

    struct DebugIdx<'a, I, Db>
    where
        I: Ingredient<Db>,
    {
        value_id: salsa::Id,
        ingredient: &'a I,
        db: PhantomData<Db>,
    }

    impl<'a, I, Db> std::fmt::Debug for DebugIdx<'a, I, Db>
    where
        I: Ingredient<Db>,
    {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            self.ingredient.fmt_index(Some(self.value_id), f)
        }
    }
```